man/systemd-measure.xml

```diff
@@ -84,29 +105,47 @@
-        the unified kernel image components from. Each option corresponds with the equally named section in
-        the unified kernel PE file. The <option>--linux=</option> switch expects the path to the ELF kernel
-        file that the unified PE kernel will wrap. All switches except <option>--linux=</option> are
-        optional. Each option may be used at most once.</para></listitem>
+        <listitem><para>When used with the <command>calculate</command> or <command>sign</command> verb,
+        configures the files to read the unified kernel image components from. Each option corresponds with
+        the equally named section in the unified kernel PE file. The <option>--linux=</option> switch expects
+        the path to the ELF kernel file that the unified PE kernel will wrap. All switches except
+        <option>--linux=</option> are optional. Each option may be used at most once.</para></listitem>
       </varlistentry>
 
       <varlistentry>
         <term><option>--current</option></term>
-        <listitem><para>When used with the <command>calculate</command> verb, takes the PCR 11 values
-        currently in effect for the system (which should typically reflect the hashes of the currently booted
-        kernel). This can be used in place of <option>--linux=</option> and the other switches listed
-        above.</para></listitem>
+        <listitem><para>When used with the <command>calculate</command> or <command>sign</command> verb,
+        takes the PCR 11 values currently in effect for the system (which should typically reflect the hashes
+        of the currently booted kernel). This can be used in place of <option>--linux=</option> and the other
+        switches listed above.</para></listitem>
       </varlistentry>
 
       <varlistentry>
         <term><option>--bank=DIGEST</option></term>
 
         <listitem><para>Controls the PCR banks to pre-calculate the PCR values for – in case
-        <command>calculate</command> is invoked –, or the banks to show in the <command>status</command>
-        output. May be used more then once to specify multiple banks. If not specified, defaults to the four
-        banks <literal>sha1</literal>, <literal>sha256</literal>, <literal>sha384</literal>,
-        <literal>sha512</literal>.</para></listitem>
+        <command>calculate</command> or <command>sign</command> is invoked –, or the banks to show in the
+        <command>status</command> output. May be used more then once to specify multiple banks. If not
+        specified, defaults to the four banks <literal>sha1</literal>, <literal>sha256</literal>,
+        <literal>sha384</literal>, <literal>sha512</literal>.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--private-key=PATH</option></term>
+        <term><option>--public-key=PATH</option></term>
+
+        <listitem><para>These switches take paths to a pair of PEM encoded RSA key files, for use with
```

Comment:
(but yeah, teaching "systemd-measure sign" to take a JSON PCR dump generated via "calculate" or "measure" as input should be easy, and is precisely what the JSON PCR dump format is good for.)

Reply:
pkcs11 would cover many cases, but not all, in many cases really only an engine is available (think of the datacenter build system, where engines take a hash, ship it off to a secure box that checks some policy from the caller and returns the signature for the hash if it passes)

Can we at least have a flag for --calculate that results in the exact same payload that `sign` would sign, so that you can do discrete signing in absence of engine support?

Reply:
well, there are pkcs11 remoting solutions, and I think that's actually a relatively nice box. But as mentioned, we can certainly add a mode where "sign" takes the output of "status" or "measure" as input. That way you can run "measure" on the build system, ship its result to a signing box, where it then does "sign", and returns the result.

Reply:
Well, so, the actual signing operation is trivial once you know what you need to sign. We use a local TPM to generate the hash to sign currently (which can be swtpm though). We could certainly stop after calculating what to sign, and generate an "incomplete signature" basically, that contains the payload to sign, but not the actual signature. This could then be shipped off to some server which "completes" it by signing the data and returning it with the signature field filled in. The JSON format is ready for that.

But I'd cross the bridge once we get there. There are lot of details to figure out: i.e. you probably want the server side to validate some things before it signs it, so that you can't make the server sign stuff is isn't supposed to sign. i.e. it should probably check in some form that what we hand it actually is a PCR 11 unified kernel blob, and nothing else. But how such verification can work if the server is "dumb" and doesn't have systemd-measure is not clear to me.

Reply:
I don't think we need to worry about policy and checks for the 'offline' mode (but we should for engine/pkcs11). In that case, it's up to the caller to do the right thing, we should just provide the content as-is.